File: python-client/cupidmq/accumulator.py
"""Batch por tamanho + timeout — espelha cupidmq producer accumulator."""

from __future__ import annotations

import time


class BatchAccumulator:
    def __init__(self, max_bytes: int, flush_timeout_secs: float) -> None:
        self._max_bytes = max(1, max_bytes)
        self._flush_timeout = max(0.001, flush_timeout_secs)
        self._items: list[bytes] = []
        self._bytes = 0
        self._opened_at: float | None = None

    def push(self, item: bytes) -> list[list[bytes]]:
        item_len = len(item)
        if item_len > self._max_bytes:
            out: list[list[bytes]] = []
            if batch := self.take_batch():
                out.append(batch)
            out.append([item])
            return out

        out = []
        if self._bytes + item_len > self._max_bytes:
            if batch := self.take_batch():
                out.append(batch)

        if self._opened_at is None:
            self._opened_at = time.monotonic()
        self._bytes += item_len
        self._items.append(item)
        return out

    def flush_if_timeout(self) -> list[bytes] | None:
        if not self._items or self._opened_at is None:
            return None
        if time.monotonic() - self._opened_at < self._flush_timeout:
            return None
        return self.take_batch()

    def take_batch(self) -> list[bytes] | None:
        if not self._items:
            return None
        self._bytes = 0
        self._opened_at = None
        batch = self._items
        self._items = []
        return batch

    @property
    def pending_items(self) -> int:
        return len(self._items)

    @property
    def pending_bytes(self) -> int:
        return self._bytes

File: master/src/producer/accumulator.rs
use bytes::Bytes;
use std::time::{Duration, Instant};

/// Accumulates payloads by size; flush on `max_bytes` or timeout.
pub struct BatchAccumulator {
    items: Vec<Bytes>,
    bytes: usize,
    max_bytes: usize,
    flush_timeout: Duration,
    opened_at: Option<Instant>,
}

impl BatchAccumulator {
    pub fn new(max_bytes: usize, flush_timeout: Duration) -> Self {
        Self {
            items: Vec::new(),
            bytes: 0,
            max_bytes: max_bytes.max(1),
            flush_timeout,
            opened_at: None,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }

    pub fn pending_bytes(&self) -> usize {
        self.bytes
    }

    pub fn pending_items(&self) -> usize {
        self.items.len()
    }

    /// Returns batches ready to send (0–2: prior flush + oversized item).
    pub fn push(&mut self, item: Bytes) -> Vec<Vec<Bytes>> {
        let item_len = item.len();
        if item_len > self.max_bytes {
            let mut out = Vec::new();
            if let Some(batch) = self.take_batch() {
                out.push(batch);
            }
            out.push(vec![item]);
            return out;
        }

        let mut out = Vec::new();
        if self.bytes.saturating_add(item_len) > self.max_bytes {
            if let Some(batch) = self.take_batch() {
                out.push(batch);
            }
        }

        if self.opened_at.is_none() {
            self.opened_at = Some(Instant::now());
        }
        self.bytes = self.bytes.saturating_add(item_len);
        self.items.push(item);
        out
    }

    pub fn flush_if_timeout(&mut self) -> Option<Vec<Bytes>> {
        if self.items.is_empty() {
            return None;
        }
        let opened = self.opened_at?;
        if opened.elapsed() < self.flush_timeout {
            return None;
        }
        self.take_batch()
    }

    pub fn take_batch(&mut self) -> Option<Vec<Bytes>> {
        if self.items.is_empty() {
            return None;
        }
        self.bytes = 0;
        self.opened_at = None;
        Some(std::mem::take(&mut self.items))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flush_on_size() {
        let mut acc = BatchAccumulator::new(10, Duration::from_secs(60));
        assert!(acc.push(Bytes::from_static(&[1; 6])).is_empty());
        let batches = acc.push(Bytes::from_static(&[2; 6]));
        assert_eq!(batches.len(), 1);
        assert_eq!(batches[0].len(), 1);
        assert_eq!(batches[0][0].len(), 6);
        assert_eq!(acc.pending_items(), 1);
    }

    #[test]
    fn oversized_item_own_batch() {
        let mut acc = BatchAccumulator::new(10, Duration::from_secs(60));
        acc.push(Bytes::from_static(&[1; 4]));
        let batches = acc.push(Bytes::from_static(&[9; 20]));
        assert_eq!(batches.len(), 2);
        assert_eq!(batches[0].len(), 1);
        assert_eq!(batches[1][0].len(), 20);
    }
}
